### PR TITLE
Add in mutations

### PR DESCRIPTION
**Pull request by ayush-goyal**
Changes
----
Adds in ability to create and edit requisitions, projects, and vendors. Can also save requisition as draft with unfilled data. Creates base admin panel. The projcts will update based on archived or not, and vendors will update based on active or not based on the settings in the admin panel. Additionally, the GraphQL cache is auto updated. Permissions for auth routes are not active yet, I will add that in in the next pull request.

### CODE REVIEW

File: frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            />
+          </Form.Item>
+          <Form.Item
+            name="archived"
+            label="Archived"
+            valuePropName="checked"
+            initialValue={initialValues ? initialValues.archived : false}
+          >
+            <Switch />
+          </Form.Item>
```
Comment (evan10s) [inline frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx:111]: I would just leave the Archived field off and set it to a sane default.  There's no (good) reason someone would create and then immediately archive a project

File: frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    );
+  }
+
+  const leadOptions = loading ? [] : data.users.map((user: any) => ({
+    label: `${user.preferredName} ${user.lastName}`,
+    value: user.id
+  }));
+
+  return (
+    <ManageContentModal
```
Comment (evan10s) [inline frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx:31]: When you create a project, click save, and then go to create another one, the old values aren't cleared in the modal.

File: frontend/src/components/requisitionForm/RequisitionForm.tsx [hunk 3]
```diff
@@ -126,7 +185,7 @@ const RequisitionForm: React.FunctionComponent = () => {
           <Col {...halfLayout}>
             <Form.Item
               name="vendor"
-              rules={[RULES.requiredRule]}
+              rules={[FORM_RULES.requiredRule]}
               label="Vendor"
             >
               <Select options={vendorOptions} showSearch optionFilterProp="label" loading={loading} />
```
Comment (evan10s) [inline frontend/src/components/requisitionForm/RequisitionForm.tsx:191]: I can just make an issue so we can address this later, but the ability to create a vendor at the same time as opening a requisition would be nice

File: frontend/src/components/requisitionForm/RequisitionForm.tsx [hunk 4]
```diff
@@ -204,7 +263,8 @@ const RequisitionForm: React.FunctionComponent = () => {
         <Row justify="center">
```
Comment (evan10s) [inline frontend/src/components/requisitionForm/RequisitionForm.tsx:263]: Would be nice if a live running total cost was shown at the bottom of the page

File: frontend/src/components/requisitionForm/RequisitionItemCard.tsx [hunk 5]
```diff
@@ -94,7 +102,7 @@ const RequisitionItemCard: React.FC<Props> = (props) => (
         <Form.Item
```
Comment (evan10s) [inline frontend/src/components/requisitionForm/RequisitionItemCard.tsx:102]: I feel like the Notes field for an individual item should be optional

File: expenses/controllers/RequisitionController.py [hunk 6]
```diff
@@ -1,19 +1,73 @@
...
+
+    @classmethod
+    def create_requisition(cls, info, data):
+        if info.context.user.has_perm("expenses.add_requisition"):
+            project = Project.objects.get(id=data["project"])
+            id_max = Requisition.objects.filter(project=project).aggregate(Max('project_requisition_id'))
+            new_data = {
+                "project": project,
+                "vendor": Vendor.objects.get(id=data["vendor"]) if "vendor" in data else None,
+                "project_requisition_id": id_max["project_requisition_id__max"] + 1,
```
Comment (evan10s) [inline expenses/controllers/RequisitionController.py:27]: This appears to break when you try to create the first requisition in a project

Steps to reproduce:

1. Create a new project
2. Fill out new requisition form
3. Click either submit button and you get the error `graphql.error.located_error.GraphQLLocatedError: unsupported operand type(s) for +: 'NoneType' and 'int'`
4. Open a new tab and go to Django's admin panel.  Create a requisition for the project
5. Go back to your requisition and submitting will work

File: frontend/src/components/requisitionForm/RequisitionItemCard.tsx [hunk 7]
```diff
@@ -81,8 +88,9 @@ const RequisitionItemCard: React.FC<Props> = (props) => (
...
           name={[props.field.name, "unitPrice"]}
           fieldKey={[props.field.fieldKey, "unitPrice"]}
-          rules={[props.rules.requiredRule, props.rules.moneyRule]}
+          rules={[FORM_RULES.requiredRule, FORM_RULES.moneyRule]}
           label="Unit Price"
+          normalize={(value: any) => parseInt(value)}
         >
           <Input prefix="$" type="number" placeholder="99.99" />
```
Comment (evan10s) [inline frontend/src/components/requisitionForm/RequisitionItemCard.tsx:95]: This field doesn't let you enter numbers with decimals

File: expenses/controllers/ProjectController.py [hunk 8]
```diff
@@ -1,17 +1,37 @@
...
+            leads = auth.get_user_model().objects.filter(id__in=data["leads"])
+            project.leads.set(leads)
+
+            return project
+
+    @classmethod
+    def update_project(cls, info, id, data):
+        if info.context.user.has_perm("expenses.change_project"):
+            query = Project.objects.filter(id=id)
+            query.update(**{k: v for k, v in data.items() if k not in ["leads"]})
```
[the remark this answers is not in the document]
Comment (ayush-goyal) [reply expenses/controllers/ProjectController.py:30]: Screen Shot 2020-07-04 at 11 37 09 PM
The graphql mutation type ProjectInput only allows these fields to be provided, so only these fields will be updated in the Project object.

File: expenses/models.py [hunk 9]
```diff
@@ -42,15 +42,15 @@ class Meta:
...
     created_by = ForeignKey(auth.get_user_model(), on_delete=models.PROTECT, related_name="created_by",
                             related_query_name="created_by")
     project = ForeignKey('Project', on_delete=models.CASCADE)
-    vendor = ForeignKey('Vendor', on_delete=models.PROTECT, limit_choices_to={"is_active": True})
+    vendor = ForeignKey('Vendor', on_delete=models.PROTECT, limit_choices_to={"is_active": True}, null=True, blank=True)
     project_requisition_id = PositiveIntegerField()
-    payment_required_by = DateTimeField()
-    other_fees = DecimalField(max_digits=15, decimal_places=4)
+    payment_required_by = DateTimeField(null=True, blank=True)
+    other_fees = DecimalField(max_digits=15, decimal_places=4, null=True, blank=True)
```
[the remark this answers is not in the document]
Comment (ayush-goyal) [reply expenses/models.py:53]: This is so when you save a requisition as a draft, the fields are empty.

File: expenses/controllers/RequisitionController.py (hunk 6 above)
Comment (ayush-goyal) [reply expenses/controllers/RequisitionController.py]: Fixed

File: frontend/src/components/requisitionForm/RequisitionItemCard.tsx (hunk 7 above)
Comment (ayush-goyal) [reply frontend/src/components/requisitionForm/RequisitionItemCard.tsx:95]: Fixed

File: expenses/controllers/RequisitionController.py [hunk 10]
```diff
@@ -1,19 +1,73 @@
...
+                "project": project,
+                "vendor": Vendor.objects.get(id=data["vendor"]) if "vendor" in data else None,
+            }
+            new_data.update({k: v for k, v in data.items() if k not in ["requisitionitemSet", "project", "vendor"]})
+
+            query.update(**new_data)
+            requisition = query.first()
+
+            # Will create new item, delete extra items in db, and update existing item respectively
+            for new_item, existing_item in zip_longest(data.requisitionitemSet, list(RequisitionItem.objects.filter(requisition=requisition))):
```
[the remark this answers is not in the document]
Comment (ayush-goyal) [reply expenses/controllers/RequisitionController.py:58]: No, the order isn't important, it takes all the data that is passed in from the graphql query, and overwrites the data based on the number of items. That way if, the items are reorganised or the info is updated, the data will be updated in the database. This is the best way I could think of, so let me know if you have any other ideas.

File: frontend/src/components/admin/manageContent/ManageContentModal.tsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            .catch((err) => {
+              hide();
+              console.error(JSON.parse(JSON.stringify(err)));
+              message.error("Error saving", 2);
+            });
+        }
+      })
+      .catch((info) => {
+        console.log("Validate Failed:", info);
+      });
```
[the remark this answers is not in the document]
Comment (ayush-goyal) [reply frontend/src/components/admin/manageContent/ManageContentModal.tsx]: Switched to async/await

File: frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx (hunk 1 above)
Comment (ayush-goyal) [reply frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx:111]: Fixed

File: frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx (hunk 2 above)
Comment (ayush-goyal) [reply frontend/src/components/admin/manageContent/formModals/ProjectFormModal.tsx:31]: Fixed

File: expenses/controllers/RequisitionController.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (evan10s) [reply expenses/controllers/RequisitionController.py:58]: What you have works so it's probably fine.  My only other idea is that you could send a partial list of updated requisition items with the IDs included and the operation (add/update/delete) and new data, but that's not really simple either

File: frontend/src/components/requisitionForm/RequisitionForm.tsx (hunk 4 above)
Comment (ayush-goyal) [reply frontend/src/components/requisitionForm/RequisitionForm.tsx:276]: Added

File: frontend/src/components/requisitionForm/RequisitionItemCard.tsx (hunk 5 above)
Comment (ayush-goyal) [reply frontend/src/components/requisitionForm/RequisitionItemCard.tsx:102]: Changed

**Pull request merged**